### PR TITLE
Feature/milestone 4

### PR DESCRIPTION
**Pull request by defoebrand**
This milestone involved:
 - The creation of a logic folder and two logic files - calculate and operate
 -  The calculate function handles the input of data into the calculator
 - The operate function handles the computing of the input calculation
 - The calculate function has been imported to App.js

### CODE REVIEW

**Comment by defoebrand** [comment]
The linter error is due to a microverse specification:
image
It is indicated to import but there is no direction on using this function at this time.

File: src/components/App.js [hunk 1]
```diff
@@ -3,10 +3,12 @@ import '../styles/App.css';
 import Display from './Display';
 import ButtonPanel from './ButtonPanel';
 
+import calculate from '../logic/calculate';
+
 function App() {
   return (
     <>
-      <Display result="calc result" />
+      <Display result="Hello!" />
```
Comment (mosaaleb) [inline src/components/App.js:11]: - [ ] We could bypass the linter issue by passing down the `calculate` function to display for now until you start working on it on the next milestone:

`<Display result="Hello" calculate={calculate} />`

File: src/logic/operate.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+  if (operation === '+') {
+    result = firstNumber.plus(secondNumber);
+  } else if (operation === '-') {
+    result = firstNumber.minus(secondNumber);
+  } else if (operation === 'x') {
+    result = firstNumber.times(secondNumber);
+  } else if (operation === '/') {
+    result = firstNumber.div(secondNumber);
+  } else if (operation === '%') {
```
Comment (mosaaleb) [inline src/logic/operate.js:16]: - [ ] `%` in our case isn't `modulus`, it's just converting a single number to a percentage

File: src/logic/operate.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import Big from 'big.js';
+
+const operate = (numberOne, numberTwo, operation) => {
+  const display = document.querySelector('.displayOutput');
```
Comment (mosaaleb) [inline src/logic/operate.js]: - [ ] You shouldn't be using `querySelector`. `operate` and `calculate` functions are merely for logic, and shouldn't get invovled in any DOM manipulation specially when you are using ReactJS. `operate` and `calculate` should only return the results.

File: src/logic/calculate.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import operate from './operate';
+
+const calculate = (calculatorData, buttonName) => {
+  const data = calculatorData;
+  const display = document.querySelector('.displayOutput');
```
Comment (mosaaleb) [inline src/logic/calculate.js]: - [ ] Same here, let's just return data `calcaultorData` object.

File: src/logic/calculate.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  if (buttonName.match(/\d|[.]/)) {
+    if (display.textContent === 'Hello!') {
+      data.total = buttonName;
```
Comment (mosaaleb) [inline src/logic/calculate.js]: - [ ] We can destruct { total, next, operation } from `caculatorData` instead.

```js
const { total, next, operation } = calculatorData;
```

**Review by mosaaleb — CHANGES_REQUESTED** [review]
#### Changes required

Well done @defoebrand :clap:. Most requirements are handled, yet there are other requirements we need to work on before moving on. Please take a look at my comments below

Please handle these requirements and submit a new code review

##### Muhammad Ebeid

**Review by mosaaleb — APPROVED** [review]
### Milestone Approved :heavy_check_mark:

Thanks, @defoebrand for implementing the past review requirements.
You can merge :+1:

- **[OPTIONAL]** There is a lot of console error from the prevoius milestones coming from `ButtonPanel` component

Screenshot from 2020-12-16 15-40-54

Let's make sure before merging to give each button a unique key. 
You can use the value since all button values are unique to assign their keys. You could also use the group key if you want to distinguish them by groups.

```js
<Button key={`${key} ${value}`} name={value} group={key} />
```

Note that this is a hard requirement but since it's not mentioned in the previous milestones, it's marked here as optional.

##### Muhammad Ebeid

**Comment by defoebrand** [comment]
Value added as key for buttons

**Pull request merged**